feat: Add Base64 global object

kriskowal opened the pull request:
This adds a global `Base64` namespace object with `encode` and `decode` functions.

phoddie commented:
Cool. More native code. ;) I will review the details in the next day or two.

One quick comment. With the text encoder/decoder and Crypto Digest integrations, I tried to establish a pattern where the code that installs the objects (for example, `modInstallTextEncoder`) is kept in the file that implements the object (textencoder.c for the previous example, and here modBase64.c). There are a few reasons for that:

- ensure the install code it is kept in up-to-date when the implementation changes
- allow other projects, beyond xsnap, to call the installation function rather than duplicating that code.

I'd prefer to do the same here, if that's OK.

dckc commented on xsnap/sources/xsnap-worker.c:
```diff
@@ -533,14 +539,24 @@ void xsBuildAgent(xsMachine* machine)
+	// Base64 = { encode, decode }
+	xsResult = xsNewObject();
+	xsVar(0) = xsNewHostFunction(xs_base64_encode, 1);
+	xsDefine(xsResult, xsID("encode"), xsVar(0), xsDontEnum);
+	xsVar(0) = xsNewHostFunction(xs_base64_decode, 1);
+	xsDefine(xsResult, xsID("decode"), xsVar(0), xsDontEnum);
+	xsDefine(xsGlobal, xsID("Base64"), xsResult, xsDontEnum);
```
why `xsDontEnum`?

kriskowal replied:
Cargo. I’ll figure out what belongs here instead.

dckc reviewed (commented):
This looks plausible.

I'll stand by for tests in agoric-sdk.

kriskowal commented:
I’m satisfied with this change, but there is a problem that we can work-around for now. The following example renders a `TypeError: Uint8Array: detached buffer` error. The error only occurs for an empty input.

https://github.com/Agoric/agoric-sdk/pull/3817/files#diff-0cb68c94fb318d0bf20d095e6203185dd74a6670c342241c28cc81e40891a824R72-R88

kriskowal commented:
> One quick comment. With the text encoder/decoder and Crypto Digest integrations, I tried to establish a pattern where the code that installs the objects (for example, `modInstallTextEncoder`) is kept in the file that implements the object (textencoder.c for the previous example, and here modBase64.c).
>
> * ensure the install code it is kept in up-to-date when the implementation changes
> * allow other projects, beyond xsnap, to call the installation function rather than duplicating that code.
> 
> I'd prefer to do the same here, if that's OK.

To that end, would you like me to carve out a PR against the module in the Moddable/XS repository?

This is the integration PR with Agoric SDK including tests https://github.com/Agoric/agoric-sdk/pull/3817

phoddie commented:
> To that end, would you like me to carve out a PR against the module in the Moddable/XS repository?

That would be perfect, thanks.

kriskowal commented:
Updated to use a builder from XS, see https://github.com/Moddable-OpenSource/moddable/pull/700

dckc commented on xsnap/makefiles/win/xsnap.mak:
```diff
@@ -30,6 +30,9 @@ C_OPTIONS = \
 	/I$(TLS_DIR) \
+	/I$(MODDABLE)/modules/data/text/decoder \
+	/I$(MODDABLE)/modules/data/text/encoder \
+	/I$(MODDABLE)/modules/data/base64 \
```
Did you happen to test the windows build? I avoid writing untested code, and I Don't Do Windows, so I have not maintained `makefiles/win`.

kriskowal commented:
I have a Windows machine but have not programmed on Windows (outside WSL)
since 2002. I’m not entirely sure where I would start, but I can look into
it.

On Sat, Sep 11, 2021 at 7:09 PM Dan Connolly ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In xsnap/makefiles/win/xsnap.mak
> <https://github.com/agoric-labs/xsnap-pub/pull/4#discussion_r706729373>:
>
> > @@ -30,6 +30,9 @@ C_OPTIONS = \
>  	/I$(PLT_DIR) \
>  	/I$(SRC_DIR) \
>  	/I$(TLS_DIR) \
> +	/I$(MODDABLE)/modules/data/text/decoder \
> +	/I$(MODDABLE)/modules/data/text/encoder \
> +	/I$(MODDABLE)/modules/data/base64 \
>
> Did you happen to test the windows build? I avoid writing untested code,
> and I Don't Do Windows, so I have not maintained makefiles/win.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/agoric-labs/xsnap-pub/pull/4#pullrequestreview-752061610>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAOXBWZ4AYDZWY7ES2KEJTUBQDV7ANCNFSM5D3H4PWA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

dckc commented:
> I have a Windows machine but have not programmed on Windows (outside WSL) since 2002. I’m not entirely sure where I would start, but I can look into it.

I would rather prune that directory or add a "abandon hope (of support) all ye who enter here" label.

kriskowal commented:
I removed the win build support in a follow-up commit, and removed the changes from the base commit.

I enabled the regression test for the degenerate base64 decode case.

I pinned https://github.com/Agoric/agoric-sdk/pull/3817 to this branch and pushed for CI.
